File: PathSumII113.java
import java.util.*;
//time complexity :  O(n)
//space complexity :  N(logN) ... N is for the LinkedList, log n is for recurssion stack
//  leet code execution : successfull
// steps: each path is computed against the targeted sum using runningSum, this was done by computing sum till the last leaf node on the left sub tree and right subtree in inorder traversal. Once running sum matches against targeted sum all the root elements in stack of deep copied(using another list) to result list.
class PathSumII113{

    class TreeNode {
        int val;
        TreeNode right;
        TreeNode left;
        TreeNode(int x){
            this.val = x;
        }
    }
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
    
        List<List<Integer>> result = new LinkedList<>();
        LinkedList<Integer> stack = new LinkedList<>();
        helper(result,stack,root,sum,0);
        return result;
    }

    private void helper(List<List<Integer>> result, LinkedList<Integer> stack, TreeNode root, int sum, int runningSum) {
    
    if(root==null) return ;

    if(root.left==null &&root.right==null){
        if(runningSum+root.val==sum){
            stack.add(root.val);
            LinkedList<Integer> trace = new LinkedList<>();
            for(int i : stack){
                trace.add(i);
            }
            result.add(trace);
            stack.removeLast();
        }
    }
    
    if(root.left!=null){
        stack.add(root.val);
        helper(result, stack, root.right, sum, runningSum+root.val);
        stack.removeLast();
    }

    if(root.right!=null){
        stack.add(root.val);
        helper(result, stack, root.left, sum, runningSum+root.val);
        stack.removeLast();
    }
    }


}

File: path-sum-ii_113.py

#Approach: Summing up path till we find leaf and backtrack to track current path
#as the pointer to list will be updating every time we will add final result into new array 
#Time Complexity O(N) for all N nodes
#Space Complexity O(H) height of tree, Because we are using list for path
#It successfully runs on the leetcode

class Solution:
    def pathSum(self, root: Optional[TreeNode], target: int) -> List[List[int]]:
        self.result=[]
        self.helper(root,target,0,[])
        return self.result
    
    def helper(self,root,target,cursum,path):
            if not root: return
            cursum+=root.val
            path.append(root.val)
            self.helper(root.left,target,cursum,(path))
            if root.left== None and root.right==None:
                if cursum==target:
                    self.result.append(list(path))
            self.helper(root.right,target,cursum,(path))
            path.pop()